### PR TITLE
feat(vtkPiecewiseWidget): Support vertical movement of points

### PR DESCRIPTION
**Pull request by bnmajor**
Vertical shift is allowed when holding control and left-click+dragging up/down. Shifting the points changes opacity.

Initial pass at fixing #188

### CODE REVIEW

**Comment by bnmajor** [comment]
@aylward @floryst I would love some feedback to make sure this is on track/meets the expectations of the issue. Thanks!

**Comment by aylward** [comment]
Looks great...but I think the opacity value is inverted...

When I raise up the opacity curve...more becomes transparent...

image

And when I lower it, less becomes transparent

image

This seems to be the opposite of what I'd expect and what the transfer function curve shows...

**Comment by bnmajor** [comment]
> Looks great...but I think the opacity value is inverted...
> 
I am sorry about that, the values are fixed now!

**Review by aylward — APPROVED** [review]
LGTM!

**Comment by aylward** [comment]
Actually - one request, if the opacity value is 0, let's not cause it to shift (that is, only apply the increase/decrease to non-zero opacity values - zero should always remain zero.  That way, the user will be able to select/threshold the background and change the non-zero opacities assigned to the data up/down without the background becoming an issue.

**Review by aylward — COMMENTED** [review]
Please see follow-on comment for one more change request - zero opacity values should always remain zero, i.e., should not be shifted up/down.

**Comment by aylward** [comment]
Weird...without changing anything, the appearance of the prostatex MRI is different in your version:

Here is the original default rendering
image


Here is the default render provided by your version
image

This version perhaps has some kind of color saturation happening?
image

**Comment by aylward** [comment]
Also - maybe you need to clip opacities to 1.0.   If I raise the global opacity really high using an all white transfer function (white is the only color in the transfer function), black stuff still appears and some odd shading wonkiness happens...

image

File: src/utils/vtk-helpers.ts [hunk 1]
```diff
@@ -128,7 +129,13 @@ export function getShiftedOpacityFromPreset(
 
     const [xmin, xmax] = effectiveRange;
     const width = xmax - xmin;
-    return points.map(([x, y]) => [(x - xmin) / width + shift, y]);
+    return points.map(([x, y]) => {
+      // Non-zero values should be affected by shift
+      // but preset values of zero should not
+      const shifted = y - shiftAlpha;
+      const yVal = y || shifted ? shifted : -1;
```
[the remark this answers is not in the document]
Comment (aylward) [reply src/utils/vtk-helpers.ts:136]: Should also clip range?

File: src/vtk/PiecewiseWidget/index.js [hunk 2]
```diff
@@ -188,7 +210,12 @@ export function extend(publicAPI, model, initialValues = {}) {
 
   vtkPiecewiseGaussianWidget.extend(publicAPI, model, initialValues);
 
-  macro.setGet(publicAPI, model, ['opacityPoints', 'opacityPointShift']);
+  macro.setGet(publicAPI, model, [
+    'opacityPoints',
+    'opacityPointShift',
+    'opacityAlphaShift',
+    'ctrlKeyIsDown',
```
Comment (floryst) [inline src/vtk/PiecewiseWidget/index.js:217]: I recommend a more descriptive name (e.g. like "shiftOpacityAlpha" or something else).

**Comment by floryst** [comment]
Sorry for ambiguity of the review. I meant to specify that `ctrlKeyIsDown` should be renamed to something more descriptive, since (1) it's not obvious what it does and (2) we don't want to imply that the ctrl key is the only key we can use for controlling the vertical/alpha opacity shift.

You can revert [9c8bc0d](https://github.com/Kitware/VolView/pull/488/commits/9c8bc0dbf0d8ad0ea39c960f8990093b3427e646). As an aside, the naming of `opacityShiftAlpha` is mostly fine, though "alpha" and "opacity" generally refer to the same concept.

**Comment by aylward** [comment]
All issues I had found have been resolved.  Works great!!

**Comment by bnmajor** [comment]
> Sorry for ambiguity of the review. I meant to specify that `ctrlKeyIsDown` should be renamed to something more descriptive, since (1) it's not obvious what it does and (2) we don't want to imply that the ctrl key is the only key we can use for controlling the vertical/alpha opacity shift.
> 
> You can revert [9c8bc0d](https://github.com/Kitware/VolView/pull/488/commits/9c8bc0dbf0d8ad0ea39c960f8990093b3427e646). 

Ah, I'm sorry for the misunderstanding there. That makes sense.

> As an aside, the naming of `opacityShiftAlpha` is mostly fine, though "alpha" and "opacity" generally refer to the same concept.

Agreed, the name is redundant. We're really just shifting the values so what about:
- `ctrlKeyIsDown` -> `shiftOpacityValues`
- `opacityAlphaShift` -> `opacityValueShift`

**Comment by aylward** [comment]
Still works great!   

This is an outstanding extension - can create exceptionally nice views quickly.

One question - why CTRL and not SHIFT?

s

**Comment by aylward** [comment]
Oops - just found a bug...switch to certain transfer functions, and this happens...

Untitled video - Made with Clipchamp

I think the transfer function is correct - but the display of that transfer function over the histogram is wrong.

**Comment by bnmajor** [comment]
> Still works great!
> 
> This is an outstanding extension - can create exceptionally nice views quickly.
> 
> One question - why CTRL and not SHIFT?
> 
> s

Great question! No reason, happy to change it to `shift` (or any key that might be preferred)

**Comment by bnmajor** [comment]
> Oops - just found a bug...switch to certain transfer functions, and this happens...
> I think the transfer function is correct - but the display of that transfer function over the histogram is wrong.

Funky... I'll check that out.

**Comment by aylward** [comment]
One more bug - seems like the transfer function being used by the LOD-reduced rendering is not being updated - only the final / high-quality transfer function is being adjusted.

**Comment by aylward** [comment]
All looks good to me!

**Comment by floryst** [comment]
> Agreed, the name is redundant. We're really just shifting the values so what about:
> 
> * `ctrlKeyIsDown` -> `shiftOpacityValues`
> * `opacityAlphaShift` -> `opacityValueShift`

These names sound good to me. Let's go with these renames.

**Comment by aylward** [comment]
FYI: Still works great!

**Comment by floryst** [comment]
LGTM!

**Pull request merged**